File: src/vrx/ros_packages/guidance_cpp/src/guidance_service.cpp
#include <memory>
#include <vector>
#include <cmath>
#include <rclcpp/rclcpp.hpp>
#include <Eigen/Dense>
#include <unsupported/Eigen/Splines>
#include <std_msgs/msg/float64_multi_array.hpp>

using Spline2d = Eigen::Spline<double, 2>;
using MatrixXd = Eigen::MatrixXd;
using VectorXd = Eigen::VectorXd;

class GuidanceService : public rclcpp::Node
{
public:
    GuidanceService() : Node("guidance_service")
    {
        publisher_ = this->create_publisher<std_msgs::msg::Float64MultiArray>("/guidance_service/output", 10);
        set_waypoints();
    }

private:
    rclcpp::Publisher<std_msgs::msg::Float64MultiArray>::SharedPtr publisher_;

    void set_waypoints()
    {
        std::vector<double> times = {0.0, 15.0, 30.0}; // Example times
        std::vector<double> north_positions = {0.0, 5.0, 10.0}; // Example north positions in meters
        std::vector<double> east_positions = {0.0, 0.0, 0.0}; // Example east positions in meters
        std::vector<double> north_velocities = {0.0, 0.0, 0.0}; // Example north velocities in meters per second
        std::vector<double> east_velocities = {0.0, 0.0, 0.0}; // Example east velocities in meters per second

        size_t n_points = times.size();

        VectorXd times_vec = Eigen::Map<VectorXd>(times.data(), n_points);
        VectorXd north_positions_vec = Eigen::Map<VectorXd>(north_positions.data(), n_points);
        VectorXd east_positions_vec = Eigen::Map<VectorXd>(east_positions.data(), n_points);
        VectorXd north_velocities_vec = Eigen::Map<VectorXd>(north_velocities.data(), n_points);
        VectorXd east_velocities_vec = Eigen::Map<VectorXd>(east_velocities.data(), n_points);

        MatrixXd waypoints(2, n_points);
        waypoints.row(0) = north_positions_vec.transpose();
        waypoints.row(1) = east_positions_vec.transpose();

        MatrixXd derivatives(2, n_points);
        derivatives.row(0) = north_velocities_vec.transpose();
        derivatives.row(1) = east_velocities_vec.transpose();

        std::vector<int> indices(n_points);
        std::iota(indices.begin(), indices.end(), 0); // Create sequence 0, 1, ..., n_points-1

        // Use Eigen::SplineFitting to fit the spline with derivatives
        auto spline = Eigen::SplineFitting<Spline2d>::InterpolateWithDerivatives(waypoints, derivatives, indices, 2);

        // Publish spline knots and coefficients
        std_msgs::msg::Float64MultiArray spline_msg;
        for (int i = 0; i < spline.knots().size(); ++i) {
            spline_msg.data.push_back(spline.knots()(i));
        }
        for (int i = 0; i < spline.ctrls().rows(); ++i) {
            for (int j = 0; j < spline.ctrls().cols(); ++j) {
                spline_msg.data.push_back(spline.ctrls()(i, j));
            }
        }
        publisher_->publish(spline_msg);

        // For debug: Evaluate splines at desired times
        VectorXd eval_times = VectorXd::LinSpaced(100, times_vec.minCoeff(), times_vec.maxCoeff());

        for (int i = 0; i < eval_times.size(); ++i)
        {
            double t = eval_times[i];
            Eigen::Matrix<double, 2, 1> eval = spline(t);
            Eigen::Matrix<double, 2, 1> eval_der = spline.derivatives(t, 1).col(1);

            RCLCPP_INFO(this->get_logger(), "Time: %.2f, N: %.2f, E: %.2f, N Vel: %.2f, E Vel: %.2f",
                        t, eval(0), eval(1), eval_der(0), eval_der(1));
        }
    }
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<GuidanceService>());
    rclcpp::shutdown();
    return 0;
}

File: src/vrx/ros_packages/guidance/guidance/guidance_service.py
import rclpy
from rclpy.node import Node
import numpy as np
from scipy.interpolate import CubicSpline
from std_msgs.msg import Float64MultiArray

class GuidanceService(Node):
    def __init__(self):
        super().__init__('guidance_service')
        self.publisher_ = self.create_publisher(Float64MultiArray, '/guidance_service/output', 10)
        self.set_waypoints()

    def set_waypoints(self):
        times = np.array([0.0, 15.0, 30.0])  # Example times
        north_positions = np.array([0.0, 5.0, 10.0])  # Example north positions in meters
        east_positions = np.array([0.0, 0.0, 0.0])  # Example east positions in meters
        north_velocities = np.array([0.0, 1/3, 0.0])  # Example north velocities in meters per second
        east_velocities = np.array([0.0, 0.0, 0.0])  # Example east velocities in meters per second

        # Fit cubic splines with specified derivatives
        spline_north = CubicSpline(times, north_positions, bc_type=((1, north_velocities[0]), (1, north_velocities[-1])))
        spline_east = CubicSpline(times, east_positions, bc_type=((1, east_velocities[0]), (1, east_velocities[-1])))

        # Evaluate splines at desired times for debugging
        eval_times = np.linspace(times.min(), times.max(), 100)
        eval_north_positions = spline_north(eval_times)
        eval_north_velocities = spline_north(eval_times, 1)
        eval_east_positions = spline_east(eval_times)
        eval_east_velocities = spline_east(eval_times, 1)

        # Log evaluated points
        for t, n, e, nv, ev in zip(eval_times, eval_north_positions, eval_east_positions, eval_north_velocities, eval_east_velocities):
            self.get_logger().info(f'Time: {t:.2f}, N: {n:.2f}, E: {e:.2f}, N Vel: {nv:.2f}, E Vel: {ev:.2f}')

        # Publish spline coefficients
        spline_msg = Float64MultiArray()
        spline_msg.data = np.concatenate([spline_north.c.ravel(), spline_east.c.ravel()]).tolist()
        self.publisher_.publish(spline_msg)

def main(args=None):
    rclpy.init(args=args)
    guidance_service = GuidanceService()
    rclpy.spin(guidance_service)
    guidance_service.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()
